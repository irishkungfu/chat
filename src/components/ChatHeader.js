import React, { Component } from "react";

const chatHeaderWrapper = {
    borderBottom: "1px black solid",
    position: "absolute",
    width: "100%",
    padding: "1rem",
    background: "white",
    textAlign: "center",
}

class ChatHeader extends Component {
    componentDidMount() {
        this.props.store.updateChatHeaderHeight(this.chatHeaderWrapper.clientHeight)
        console.log(this.props.store.chatHeaderHeight)

    }
    render() {
        return (
            <div
                ref={(chatHeaderWrapper) => this.chatHeaderWrapper = chatHeaderWrapper}
                style={chatHeaderWrapper}
            >
                Chat Tron 2000
                </div>
        )
    }
}


export default ChatHeader