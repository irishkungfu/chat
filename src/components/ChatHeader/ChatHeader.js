import React, { Component } from "react";
import styled from 'styled-components';
import { observer } from "mobx-react";
import { func, shape } from 'prop-types'


const ChatHeaderWrapper = styled.div`
    border-bottom: 1px black solid;
    position: absolute;
    width: 100%;
    padding: 1rem;
    background: white;
    text-align: center;
`

@observer

class ChatHeader extends Component {
    static propTypes = {
        store: shape({
            updateHeight: func,
        }).isRequired,
    }
    componentDidMount() {
        this.props.store.updateHeight(this.headerWrapper.clientHeight, "header")
    }
    render() {
        return (
            <ChatHeaderWrapper
                ref={(headerWrapper) => this.headerWrapper = headerWrapper}
            >
                Chat Tron 2000
                </ChatHeaderWrapper>
        )
    }
}


export default ChatHeader