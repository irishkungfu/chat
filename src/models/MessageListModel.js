
import { observable, action, flow } from "mobx"


import MessageModel from "./MessageModel"

export default class MessageListModel {

    @observable messages = [];
    @observable newMessage = ""
    @observable awaitingMessage = "done"
    @observable chatInputHeight = 50 // base value, calculated value will override
    @observable chatHeaderHeight = 50 // base value, calculated value will override

    /**
     * Gets the calculated height and sets required variable
     * @param {int} newHeight 
     * @param {string} element expects "header" or "input"
     */
    @action
    updateHeight(newHeight, element) {
        switch (element) {
            case "header":
                this.chatHeaderHeight = newHeight
                break
            case "input":
                this.chatInputHeight = newHeight
                break
            default:
                console.warn("No Argument added for element in updateHeight")
        }

    }
    /**
     * adds a message to state's message array
     * @param {string} messageBody 
     * @param {bool} isSender 
     * @param {string} messageType // text, component, list - default text
     */
    @action
    addMessage(messageBody, isSender = false, messageType = "text") {
        this.messages.push(new MessageModel(messageBody, isSender, messageType))
    }
    @action
    clearMessage() {
        this.newMessage = "";
    }

    @action
    sendMessage = flow(function* (messageBody, messageType) {
        this.addMessage(messageBody, true, messageType)
        this.clearMessage()
        try {
            const res = yield this.fetchNewMessage(Math.floor(Math.random() * 50) + 1)
            let data = yield res.json();
            return data
        } catch (error) {
            this.awaitingMessage = "error"
            return error
        }
    })
    /**
     * fetch messages async call
     * @param {int} id 
     */
    @action
    fetchNewMessage = flow(function* (id) {
        this.awaitingMessage = "pending"

        try {
            const res = yield fetch(`http://localhost:3000/messages/${id}`)
            let data = yield res.json();
            this.addMessage(data.messageBody, false, data.messageType)
            this.awaitingMessage = "done"
            return data
        } catch (error) {
            this.awaitingMessage = "error"
            return error
        }
    })
}