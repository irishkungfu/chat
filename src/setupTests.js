import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

export function configureAdapter() {
    configure({ adapter: new Adapter() });
}