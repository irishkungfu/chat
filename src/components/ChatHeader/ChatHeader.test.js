import React from 'react';
import ChatHeader from './ChatHeader';
import { shallow } from 'enzyme';
import { configureAdapter } from "../../setupTests"
import MessageListModel from "../../models/MessageListModel"
configureAdapter()

describe('ChatHeader', () => {
    const store = new MessageListModel()
    let wrapper;
    beforeEach(() => wrapper = shallow(<ChatHeader store={store} />));

    it('should render correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })
})