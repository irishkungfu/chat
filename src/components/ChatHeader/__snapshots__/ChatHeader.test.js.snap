// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatHeader should render correctly 1`] = `undefined`;

exports[`ChatWrapper should render correctly 1`] = `undefined`;
